rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+insert into bucket (
+                name,
+                policy_document, -- nullable if policy_id is provided
+                policy_id -- nullable if policy_document is provided
+            ) values (
+                      'some_bucket',
+                      NULL,
+                      iasql_create_policy( -- returns policy ID after creation
+                          'some_policy_name',
+                          '2012-10-17', -- nullable, defaults to 2012-10-17
+                          iasql_create_policy_statement -- returns policy statement ID after creation
+                            '<policy_id>',
+                            'allow',
+                            'aws:arn:123456789012:bucket-resource',
+                            's3:GetObjects', -- comma separated list of actions
+                            '123456789012' -- comma separated list of principals
+                            ), -- Statement 1
+                          iasql_create_policy_statement(...), -- Statement 2
+                          ...
+                          iasql_create_policy_statement(...) -- Statement N
+                    );
+```
+
+### ER diagram for s3 example
+![Proposed model for s3](./assets/004/s3_example.png)
+
+And for Updating, we can equally have something like this:
+
+```sql
+UPDATE bucket set policy_id = iasql_create_policy( -- returns policy ID after creation
+                              'some_policy_name',
+                              '2012-10-17', -- nullable, defaults to 2012-10-17
+                              iasql_create_policy_statement -- returns policy statement ID after creation
+                                '<policy_id>',
+                                'allow',
+                                'aws:arn:123456789012:bucket-resource',
+                                's3:GetObjects', -- comma separated list of actions
+                                '123456789012' -- comma separated list of principals
+                              ), -- Statement 1
+                              iasql_create_policy_statement(...), -- Statement 2
+                              ...
+                              iasql_create_policy_statement(...) -- Statement N
+                        ) WHERE id = '<bucket_id>';
+```
+
+These functions can be used to create policies for any AWS service (SQS, VPC, SNS etc.). We can also add more functions to create policies for
+services in other clouds like GCP, Azure (out of scope).
+
+Now, once the JSON blob is generated from the `Policy` and `Statement` tables, the data in these tables are no longer needed.
+We can choose to keep them but the tables will keep growing and will span billions of records in no time. I recommend we
+have a retention policy for the tables (perhaps we can retain them for 12-24 hours and then delete).
+We can also choose to delete them immediately after the JSON blob is generated, though a periodic bulk-delete is cheaper.
```

Comment:
hmm but if we delete the policy, how is the relation between tabled kept? Also, if removed, how are we going to be able to know if there was a change in the policy for a resource while doing `sync` or `apply`?

Reply:
The intention is to only use the tables for input, just to generate the JSON blob. Once the JSON blob is generated and stored on the entity, we can discard the table records and do all the validations and checks for changes in the same way we currently do now (using the blobs).